smarts/core/vehicle.py

```diff
@@ -48,18 +48,23 @@
 from .utils.math import rotate_around_point
 
 
-@dataclass(frozen=True)
+@dataclass
```

Comment:
Because I wanted to use the `updated` field to track whether a vehicle state had already been used in a pybullet update.  See around line 629 of `smarts.py` and consider the `step()` method in `ExternalProvider.py`.   Something like this is necessary because it's now possible that a `VehicleState` is sent from the external provider and gets used in a physics update on a step at time `t`, but then there's no update from the external provider at time `t + 1` (because its clock is out of sync).  In that case, we don't want to re-send the same vehicle state to pybullet.

Honestly, all of this was a bit of a gamble in order to try to avoid adding locks to `ExternalProvider`.    (I wanted `state_update()` there to be as "atomic" as possible.  It's still got a hazard, but I talked myself into thinking it was ok.)  However, I'm not yet _fully_ convinced about whether we can actually avoid locks there.  If we end up adding locks, then we might be able to revert the "updated" change to `VehicleState`.   (... although I _just_ added the `set_privileged()` method to it too as part of this review, but that should be easy enough to revert.) 

In the meantime, I'll at least add a comment about this in `ExternalProvider.py`.

Reply:
Because I wanted to use the updated field to track whether a vehicle state had already been used in a pybullet update. See around line 629 of smarts.py and consider the step() method in ExternalProvider.py. Something like this is necessary because it's now possible that a VehicleState is sent from the external provider and gets used in a physics update on a step at time t, but then there's no update from the external provider at time t + 1 (because its clock is out of sync). In that case, we don't want to re-send the same vehicle state to pybullet.